src/Compilers/CSharp/Test/Semantic/SourceGeneration/SyntaxAwareGeneratorTests.cs

```diff
@@ -2077,6 +2077,66 @@ void Function()
             Assert.Equal("Simulated cancellation from external source", results.Results[0].Exception!.Message);
         }
 
+        [Fact]
+        public void Syntax_Provider_Doesnt_Attribute_Incorrect_Timing()
+        {
+
+            var source = @"
+class C 
+{
+    int Property { get; set; }
+
+    void Function()
+    {
+        var x = 5;
+        x += 4;
+    }
+}
+";
+            var parseOptions = TestOptions.Regular.WithLanguageVersion(LanguageVersion.Preview);
+            Compilation compilation = CreateCompilation(source, options: TestOptions.DebugDll, parseOptions: parseOptions);
+            compilation.VerifyDiagnostics();
+
+            Assert.Single(compilation.SyntaxTrees);
+
+            var sleepTimeInMs = 50;
+            var testGenerator = new PipelineCallbackGenerator(ctx =>
+            {
+                ctx.RegisterSourceOutput(ctx.SyntaxProvider.CreateSyntaxProvider<object>((s, _) => s is AssignmentExpressionSyntax, (c, _) => { Thread.Sleep(sleepTimeInMs); return true; }), (spc, s) => { });
+            }).AsSourceGenerator();
+
+            var testGenerator2 = new PipelineCallbackGenerator2(ctx =>
+            {
+                ctx.RegisterSourceOutput(ctx.SyntaxProvider.CreateSyntaxProvider<object>((s, _) => s is AssignmentExpressionSyntax, (c, _) => { Thread.Sleep(sleepTimeInMs); return true; }), (spc, s) => { });
+                ctx.RegisterSourceOutput(ctx.SyntaxProvider.CreateSyntaxProvider<object>((s, _) => s is AssignmentExpressionSyntax, (c, _) => { Thread.Sleep(sleepTimeInMs); return true; }), (spc, s) => { });
+            }).AsSourceGenerator();
+
+            GeneratorDriver driver = CSharpGeneratorDriver.Create(new[] { testGenerator, testGenerator2 }, parseOptions: parseOptions);
+            driver = driver.RunGenerators(compilation);
+
+            var timing = driver.GetTimingInfo();
+
+            Assert.NotEqual(TimeSpan.Zero, timing.ElapsedTime);
+            Assert.Equal(2, timing.GeneratorTimes.Length);
+
+            // check generator one took at least 'sleepTimeInMs'
```

Comment:
Is there a guarantee that Thread.Sleep(50) ensures that underlying timer will advance far enough? I'm always paranoid with mixing different types of timers. The Win32 Sleep call at least:

> The system clock "ticks" at a constant rate. If dwMilliseconds is less than the resolution of the system clock, the thread may sleep for less than the specified length of time. If dwMilliseconds is greater than one tick but less than two, the wait can be anywhere between one and two ticks, and so on.

Reply:
Yeah, I agonized over this as I _really_ don't like relying on timing information, but couldn't really think of a better way to test it. Anecdotally 50ms seemed a good balance between not slowing down tests too much, but always passing.

Reply:
Did you consider the following pattern: 

```
using var mre = new ManualResetEvent(initialState: false);
...
var testGenerator = new PipelineCallbackGenerator(ctx =>
            {
                ctx.RegisterSourceOutput(ctx.SyntaxProvider.CreateSyntaxProvider<object>((s, _) => s is AssignmentExpressionSyntax, (c, _) => { mre.WaitOne(); return true; }), (spc, s) => { });
            }).AsSourceGenerator();
  
...
while (!<did timer elapse fare enough>)
{
  Thread.Sleep(1);
}
mre.Set();
```